--- a/net/minecraft/world/entity/monster/EntitySkeletonAbstract.java
+++ b/net/minecraft/world/entity/monster/EntitySkeletonAbstract.java
@@ -45,6 +45,11 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.IBlockData;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityArmorChangeEvent;
+// CraftBukkit end
+
 public abstract class EntitySkeletonAbstract extends EntityMonster implements IRangedEntity {
 
     private final PathfinderGoalBowShoot<EntitySkeletonAbstract> bowGoal = new PathfinderGoalBowShoot<>(this, 1.0D, 20, 15.0F);
@@ -108,6 +113,7 @@
                 if (itemstack.f()) {
                     itemstack.setDamage(itemstack.getDamage() + this.random.nextInt(2));
                     if (itemstack.getDamage() >= itemstack.i()) {
+                        CraftEventFactory.callEntityEquipArmorEvent(this, itemstack, ItemStack.EMPTY, EntityArmorChangeEvent.ChangeReason.BREAK, EnumItemSlot.HEAD); // CraftBukkit
                         this.broadcastItemBreak(EnumItemSlot.HEAD);
                         this.setSlot(EnumItemSlot.HEAD, ItemStack.EMPTY);
                     }
@@ -195,8 +201,19 @@
         double d3 = Math.sqrt(d0 * d0 + d2 * d2);
 
         entityarrow.shoot(d0, d1 + d3 * 0.20000000298023224D, d2, 1.6F, (float) (14 - this.level.getDifficulty().a() * 4));
+        // CraftBukkit start
+        org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(this, this.getItemInMainHand(), null, entityarrow, net.minecraft.world.EnumHand.MAIN_HAND, 0.8F, true);
+        if (event.isCancelled()) {
+            event.getProjectile().remove();
+            return;
+        }
+
+        if (event.getProjectile() == entityarrow.getBukkitEntity()) {
+            level.addEntity(entityarrow);
+        }
+        // CraftBukkit end
         this.playSound(SoundEffects.SKELETON_SHOOT, 1.0F, 1.0F / (this.getRandom().nextFloat() * 0.4F + 0.8F));
-        this.level.addEntity(entityarrow);
+        // this.world.addEntity(entityarrow); // CraftBukkit - moved up
     }
 
     protected EntityArrow b(ItemStack itemstack, float f) {
